### ISSUE TITLE
Setting non-zero `num_workers` in DataLoader is a problem

### ISSUE THREAD

**Issue by sarthakpati** [CBICA/GaNDLF#218]
After multiple failed runs in the last 2-3 weeks for segmentation training with @iskobleva, I came across this beautiful post on PyTorch GH: https://github.com/pytorch/pytorch/issues/13246

TLDR: PyTorch training (regardless of whether running on CPU or GPU) explodes RAM requirement because of the way Python manages multiprocessing (an excellent explanation with possible solutions has been presented in [a comment](https://github.com/pytorch/pytorch/issues/13246#issuecomment-905703662))

This brings us to the question on **how do we handle this for GaNDLF**? There are a few options (again, look at [the comment summarizing the issue](https://github.com/pytorch/pytorch/issues/13246#issuecomment-905703662) for other solutions):

1. Rewrite queueing mechanism by bypassing torchio - not desirable since we would need to make substantial code changes 
2. [Set up shared memory within the script](https://github.com/pytorch/pytorch/issues/13246#issuecomment-487042977): need to check with CUBIC sys-admins for this
3. Ensure that `num_workers` is always set to `0` for training - quickest solution to get us training, but quite possibly makes the training process slower

**Comment by sarthakpati** [comment]
Note that reverting #213 seems to get training to work, but this has logic issues which prevents `one_hot` and `reverse_one_hot` from working correctly, which is why the training only _seemed_ to be going on without memory issue.

**Comment by sarthakpati** [comment]
Added a single line for this check, while we debug the issue:

https://github.com/CBICA/GaNDLF/blob/d0f4a85f24d55e3fd35da6175083f753ab656cfa/GANDLF/compute/training_loop.py#L189

```python
    if params["q_num_workers"] > 0:
        print(
            "\n\n********\nWARNING: Setting 'num_workers' > 0 will causes unexpected memory issues; see https://github.com/CBICA/GaNDLF/issues/218 \n********\n\n"
        )
```

**Comment by sarthakpati** [comment]
Tracking garbage collection here: https://github.com/sarthakpati/GaNDLF/tree/218_memory-debug

**Comment by sarthakpati** [comment]
Update as of 2021/10/05::1520:

After speaking with Mark, I realized that the lazy loading was getting activated, which get me to take a look at the Dataloader more closely, and I realized that the RGB check for the label was leading to the **ENTIRE DATA** getting loaded. After removing it as of https://github.com/sarthakpati/GaNDLF/commit/ce051ee9bd0646e979c0cb6d90fdb2df29b37687, the memory on the cluster is already significantly better, even on high `num_workers`.
